### PR TITLE
first pass at auditing the code.

### PR DESCRIPTION
**Pull request by olegzil**
## Description
First pass at auditing the SDK Kotlin code. 
Removed "!!" in instance. In other places replaced "?." with ifNotNull(). Replaced "if" with "when" in a few cases.

### Goal
Code cleanup 

## Testing

Please describe the manual tests that you ran to verify your changes

- [x ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [ x] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [x ] My code follows the style guidelines of this project
- [x ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt [hunk 1]
```diff
@@ -219,20 +219,15 @@ class RouteUtils {
...
-    ): BannerInstructions? {
-        currentStep?.bannerInstructions()?.let {
+    ) = currentStep?.bannerInstructions()?.let {
             val instructions: List<BannerInstructions> = sortBannerInstructions(it)
-            for (instruction in instructions) {
-                val distanceAlongGeometry = instruction.distanceAlongGeometry().toInt()
-                if (distanceAlongGeometry >= stepDistanceRemaining.toInt()) {
-                    return instruction
-                }
+            instructions.firstOrNull { instruction ->
```
Comment (Lebedinsky) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt]: Are you sure it is the equal with `for` looping by instructions? I mean your implementation changes the logic of this method. If this new logic is correct - super.... but if not ->... )

File: app/src/main/java/com/mapbox/services/android/navigation/testapp/example/ui/ExamplePresenter.kt [hunk 2]
```diff
@@ -121,8 +121,8 @@ class ExamplePresenter(private val view: ExampleView, private val viewModel: Exa
     }
 
     fun onDestinationFound(feature: CarmenFeature) {
-        feature.center()?.let {
-            viewModel.destination.value = it
+        feature.center()?.let { point ->
```
Comment (Lebedinsky) [inline app/src/main/java/com/mapbox/services/android/navigation/testapp/example/ui/ExamplePresenter.kt]: 👍

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt [hunk 3]
```diff
@@ -219,20 +219,15 @@ class RouteUtils {
     fun findCurrentBannerInstructions(
         currentStep: LegStep?,
         stepDistanceRemaining: Double
-    ): BannerInstructions? {
-        currentStep?.bannerInstructions()?.let {
+    ) = currentStep?.bannerInstructions()?.let {
```
Comment (RingerJK) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt]: return value is not obvious 
```suggestion
    ): BannerInstructions? = currentStep?.bannerInstructions()?.let {
```

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt [hunk 4]
```diff
@@ -219,20 +219,15 @@ class RouteUtils {
...
             val instructions: List<BannerInstructions> = sortBannerInstructions(it)
-            for (instruction in instructions) {
-                val distanceAlongGeometry = instruction.distanceAlongGeometry().toInt()
-                if (distanceAlongGeometry >= stepDistanceRemaining.toInt()) {
-                    return instruction
-                }
+            instructions.firstOrNull { instruction ->
+                instruction.distanceAlongGeometry().toInt() >= stepDistanceRemaining.toInt()
```
Comment (RingerJK) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt]: looks weird, not sure that it's equal to previous approach 🤔

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt (hunk 1 above)
Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt]: Is this intentional @olegzil? It doesn't seem this is behaving as before. Overall implementation is different. Could you confirm?

Comment (olegzil) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt]: The logic has not changed. This is the sudo-code that describes the logic as it was implemented in the original code and remains the same in the modified code:
if currentStep?.bannerInstructions()? != null
*       generate a sorted BannerInstructions list
*       find a distanceAlongGeometry >= stepDistanceRemaining
*       if found return the banner instruction associated with the distance
*       else if the sorted instruction list is not empty return the first instruction
*       else return null
In short, find the banner instruction that contains a distance less then or equal to the step distance remaining. If one is found return that banner instruction. If not found, then if the banner list is null, return null, otherwise return the first instruction.
*       return null if the banner instruction list is null
*

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt (hunk 4 above)
Comment (olegzil) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt]: Identical to previous logic

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt (hunk 3 above)
Comment (olegzil) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt]: 👍

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt (hunk 1 above)
Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt]:
```kotlin
inline fun <T> Iterable<T>.firstOrNull(
    predicate: (T) -> Boolean
): T?
```

It's doing the same 👌 Look at the implementation 👀 

```kotlin
public inline fun <T> Iterable<T>.firstOrNull(predicate: (T) -> Boolean): T? {
    for (element in this) if (predicate(element)) return element
    return null
}
```

:shipit: @olegzil 
 
cc @Lebedinsky @RingerJK

**Comment by Guardiola31337** [comment]
@olegzil CI is ❌ because of `ktlint` 👀 

```
> Task :app:ktlint

> Task :libandroid-navigation:ktlint
/root/code/libandroid-navigation/src/main/java/com/******/services/android/navigation/v5/utils/RouteUtils.kt:222:7: Unexpected spacing before ":"

> Task :libandroid-navigation:ktlint FAILED
```

Other than that this looks good to me 🚀
